apps/accounts/views.py

```diff
@@ -12,24 +13,37 @@
 from accounts.serializers import UserProfileSerializer
 
 
+def find_user(identity):
+    try:
+        return UserProfile.objects.get(
+            Q(fxa_id=identity['uid']) | Q(email=identity['email']))
+    except UserProfile.DoesNotExist:
+        return None
+    except UserProfile.MultipleObjectsReturned:
+        # This shouldn't happen, so let it raise.
+        raise
+
+
 class LoginView(APIView):
 
     @waffle_switch('fxa-auth')
     def post(self, request):
         if 'code' not in request.DATA:
-            return Response({'error': 'No code provided.'}, status=400)
+            return Response({'error': 'No code provided.'}, status=422)
 
         try:
             identity = verify.fxa_identify(request.DATA['code'],
                                            config=settings.FXA_CONFIG)
         except verify.IdentificationError:
             return Response({'error': 'Profile not found.'}, status=401)
-        try:
-            user = UserProfile.objects.get(email=identity['email'])
-        except UserProfile.DoesNotExist:
-            return Response({'error': 'User does not exist.'}, status=400)
+        user = find_user(identity)
+        if user is None:
+            return Response({'error': 'User does not exist.'}, status=422)
         else:
-            login(request._request, user)
+            if (user.fxa_id != identity['uid'] or
+                    user.email != identity['email']):
+                user.update(fxa_id=identity['uid'], email=identity['email'])
```

Comment:
If the user has never logged in with FxA then we need to find the account based on the email. But if they have logged in with FxA we should find it using the user ID that FxA has in case they start allowing changing of emails.

This is really just to set the `fxa_id` but at some point the user may be able to change their email address when signing up with FxA so it should copy that over too.

I'll add a docstring to `find_user`.